### ISSUE TITLE
libtxt: Mixed Text (RTL and LTR) breaks the RTL justification

### ISSUE THREAD

**Issue by OsamaAbbas** [flutter/flutter#44449]
As per https://github.com/flutter/flutter/issues/28700#issuecomment-532884518 , I'm openning a new issue.

For a detailed explanation of the issue and how to reproduce the problematic behavior: please refer to this https://github.com/flutter/flutter/issues/28700#issuecomment-528906329 and https://github.com/flutter/flutter/issues/28700#issuecomment-533181260 .

Related Issues (but each has different aspects as well):

* https://github.com/flutter/flutter/issues/34610
* https://github.com/flutter/flutter/issues/39755

I'm about to release my product soon. It'd be great if this issue could be fixed.

**Comment by iapicca** [comment]
Hi @OsamaAbbas 
can you please provide more details about 
how the current issue differs from the open issues [ [1](https://github.com/flutter/flutter/issues/39755), [2](https://github.com/flutter/flutter/issues/34610) ] you linked above
thank you

**Comment by OsamaAbbas** [comment]
The current issue (as well as its parent https://github.com/flutter/flutter/issues/28700) is concerned with the RTL justification. Almost it is fixed in #28700, but the remaining part is that when we have a mixed text (i.e. with any digits or numbers or latin text), the RTL justification breaks.

The other two issues deals with other problems that happens with mixed text, not the justification issue itself. For example, #39755 is about selection inaccuracy (which I have partially solved by using a custom font).

They are related because the reason that causes them might be the same (and might be not).

I only opened the current issue because @Hixie [said so](https://github.com/flutter/flutter/issues/28700#issuecomment-532884518). Otherwise the issue was being discussed in #28700.

**Comment by OsamaAbbas** [comment]
I just found that the issue occurs also if we use `TextDirection.ltr`! So the justification is broken in case of mixed RTL and LTR characters, whether the context text is RTL or LTR.

![image](https://user-images.githubusercontent.com/3346502/69526412-78488980-0fad-11ea-86dd-078520612d74.png)
```Dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final String mixedText =
      "1 - العربية لغة رسمية في كل دول الوطن العربي إضافة إلى كونها لغة رسمية في تشاد وإريتريا وإسرائيل. وهي إحدى اللغات الرسمية الست في منظمة الأمم المتحدة، ويُحتفل باليوم العالمي للغة العربية في ١٨ ديسمبر كذكرى اعتماد العربية بين لغات العمل في الأمم المتحدة.\n1 - العربية لغة رسمية في كل دول الوطن العربي إضافة إلى كونها لغة رسمية في تشاد وإريتريا وإسرائيل. وهي إحدى اللغات الرسمية الست في منظمة الأمم المتحدة، ويُحتفل باليوم العالمي للغة العربية في ١٨ ديسمبر كذكرى اعتماد العربية بين لغات العمل في الأمم المتحدة.\n";

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        appBar: AppBar(),
        body: Center(
          child: Padding(
            padding: EdgeInsets.all(20.0),
            child: Text.rich(
              TextSpan(
                text: mixedText,
                style: TextStyle(
                  color: Colors.black,
                ),
              ),
              textDirection: TextDirection.ltr,
              textAlign: TextAlign.justify,
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by saioufe** [comment]
+1

**Comment by saioufe** [comment]
the numbers will Always break the line

**Comment by GaryQian** [comment]
A good repro is posted in https://github.com/flutter/flutter/issues/48335

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.22.0-2.0.pre.93`.

<img src="https://user-images.githubusercontent.com/68219924/91431232-c2c2a380-e868-11ea-9720-5d1cd1779f29.png" height="600" />

<details>
 <summary>Minimal reproducible code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Sample',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Sample'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Container(
                color: Colors.grey,
                height: 200,
                width: 300,
                child: Text(
                  "123132ميزة أو فائدة؟ ولكن من لديه الحق أن ينتقد شخص ما أراد أن يشعر بالسعادة التي لا تشوبها عواقب أليمة أو آخر أراد أن يتجنب الألم الذي ربما تنجم عنه بعض المتعة ؟ علي الجانب الآخر نشجب ون53412512ستنكر هؤلاء الرجال المفتونون بنشوة اللحظة الهائمون في رغباتهم فلا يدركون م123123ا يعقبها من الألم والأسي المحتم، واللوم كذلك يشمل هؤلاء الذين أخفقوا في واجباتهم نتيجة لضعف إرادتهم فيت213132ساوي مع هؤلاء الذين يتجنبون وينأون عن تحمل الكدح",
                  textAlign: TextAlign.justify,
                  textDirection: TextDirection.rtl,
                ),
              ),
              SizedBox(
                height: 20,
              ),
              Container(
                color: Colors.grey,
                height: 200,
                width: 300,
                child: Text(
                  "ميزة أو فائدة؟ ولكن من لديه الحق أن ينتقد شخص ما أراد أن يشعر بالسعادة التي لا تشوبها عواقب أليمة أو آخر أراد أن يتجنب الألم الذي ربما تنجم عنه بعض المتعة ؟ علي الجانب الآخر نشجب ونستنكر هؤلاء الرجال المفتونون بنشوة اللحظة الهائمون في رغباتهم فلا يدركون ما يعقبها من الألم والأسي المحتم، واللوم كذلك يشمل هؤلاء الذين أخفقوا في واجباتهم نتيجة لضعف إرادتهم فيتساوي مع هؤلاء الذين يتجنبون وينأون عن تحمل الكدح",
                  textAlign: TextAlign.justify,
                  textDirection: TextDirection.rtl,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```
</details>
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 1.22.0-2.0.pre.93, on Microsoft Windows [Version 10.0.19041.450], locale en-US)
    • Flutter version 1.22.0-2.0.pre.93 at C:\Development\flutter_master
    • Framework revision f35a8b7470 (4 hours ago), 2020-08-26 21:33:03 -0700
    • Engine revision 388193a67b
    • Dart version 2.10.0 (build 2.10.0-62.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at C:\Users\marku\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30320.27
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (4 available)
    • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)           • windows       • windows-x64    • Microsoft Windows [Version 10.0.19041.450]
    • Web Server (web)            • web-server    • web-javascript • Flutter Tools
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 84.0.4147.135

• No issues found!
```
</details>

**Comment by JohannaShiri** [comment]
Having also this problem in Hebrew. Any estimation when it should be fixed?

**Comment by mohammadamir98** [comment]
Hi @OsamaAbbas 
Any solutions found on this yet? I am using RichText with Textspans inside Listview, so using text is not an option for me.

**Comment by OsamaAbbas** [comment]
Check my workaround here: https://github.com/flutter/flutter/issues/28700#issuecomment-558230313 .

**Comment by JohannaShiri** [comment]
> Check my workaround here: [#28700 (comment)](https://github.com/flutter/flutter/issues/28700#issuecomment-558230313) .

I need the ability to copy the text :( so that would not be suitable for me...

**Comment by mhorizon** [comment]
i have same problem 
when using widget span in arabic text line
widget spans showed ltr but text is rtl!!!
please help me if there is any solution!!!

**Comment by KotlinByte** [comment]
same problem on Flutter (Channel stable, 2.5.1)

**Comment by OsamaAbbas** [comment]
According to my (very quick) test: this issue is solved in v3.7.0. Mixed text is now correctly justified, whether the context is `TextDirection.ltr` or `TextDirection.rtl`. I used the simple example in [this comment](https://github.com/flutter/flutter/issues/44449#issuecomment-558057639) with small modifications.

![image](https://user-images.githubusercontent.com/3346502/214937412-3bbe4380-1a87-4700-bf66-73d2fdfeea98.png)

This issue could be closed now.

**Comment by darshankawar** [comment]
Thanks for confirming. Closing this as fixed per above comment.
